integration-tests/tests/client/sharding_upgrade.rs

```diff
@@ -9,21 +7,202 @@ use near_logger_utils::init_test_logger;
+struct TestShardUpgradeEnv {
+    env: TestEnv,
+    initial_accounts: Vec<AccountId>,
+    init_txs: Vec<SignedTransaction>,
+    txs_by_height: HashMap<u64, Vec<SignedTransaction>>,
+    #[allow(unused)]
+    epoch_length: u64,
+    num_validators: usize,
+    num_clients: usize,
+}
+
+/// Test shard layout upgrade. This function runs `env` to produce and process blocks
+/// from 1 to 3 * epoch_length + 1, ie, to the beginning of epoch 3.
+/// Epoch 0: 1 shard
+/// Epoch 1: 1 shard, state split happens
+/// Epoch 2: shard layout upgrades to simple_night_shade_shard,
+impl TestShardUpgradeEnv {
+    fn new(
+        epoch_length: u64,
+        num_validators: usize,
+        num_clients: usize,
+        num_init_accounts: usize,
+        gas_limit: Option<u64>,
+    ) -> Self {
+        let mut rng = thread_rng();
+        let validators: Vec<AccountId> = (0..num_validators)
+            .map(|i| format!("test{}", i).to_string().parse().unwrap())
+            .collect();
+        let initial_accounts =
+            [validators, gen_unique_accounts(&mut rng, num_init_accounts)].concat();
+        let genesis =
+            setup_genesis(epoch_length, num_validators as u64, initial_accounts.clone(), gas_limit);
+        let chain_genesis = ChainGenesis::from(&genesis);
+        let env = TestEnv::builder(chain_genesis)
+            .clients_count(num_clients)
+            .validator_seats(num_validators)
+            .runtime_adapters(create_nightshade_runtimes(&genesis, num_clients))
+            .build();
+        Self {
+            env,
+            initial_accounts,
+            epoch_length,
+            num_validators,
+            num_clients,
+            init_txs: vec![],
+            txs_by_height: HashMap::new(),
+        }
+    }
+
+    /// `init_txs` are added before any block is produced
+    fn set_init_tx(&mut self, init_txs: Vec<SignedTransaction>) {
+        self.init_txs = init_txs;
+    }
+
+    /// `txs_by_height` is a hashmap from block height to transactions to be included at block at
+    /// that height
+    fn set_tx_at_height(&mut self, height: u64, txs: Vec<SignedTransaction>) {
+        self.txs_by_height.insert(height, txs);
+    }
+
+    /// produces and processes the next block
+    /// also checks that all accounts in initial_accounts are intact
+    fn step(&mut self) {
+        let env = &mut self.env;
+        let head = env.clients[0].chain.head().unwrap();
+        let height = head.height + 1;
+
+        // add transactions for the next block
+        if height == 1 {
+            for tx in self.init_txs.iter() {
+                for j in 0..self.num_validators {
+                    env.clients[j].process_tx(tx.clone(), false, false);
+                }
+            }
+        }
+
+        // chunks for the block at height has already been produced at this point, so we need to
+        // add these transactions at i+1 for them to be included in block at height i
```

Comment:
do you mean block at height i + 1?

Reply:
I modified the comments, hopefully it's more clear.